### PR TITLE
Allow Python dynamic providers to capture secrets

### LINKED ISSUE #15539: Provide a way to opt-out of secretness for Python dynamic providers
The serialized Python dynamic provider code is marked as a secret by default with https://github.com/pulumi/pulumi/pull/13315. This was done to ensure any credentials used by a dynamic provider would not be stored in plaintext in the state. However, this can lead to performance issues due to https://github.com/pulumi/pulumi/issues/15538.

We should consider providing a way to opt-out of secretness, or even consider changing the default back and making it opt-in. Perhaps we could come up with a way to detect secretness similar to what the [Node.js SDK does](https://github.com/pulumi/pulumi/pull/13329) for its dynamic provider implementation.

In the meantime, if you know your Python dynamic provider implementation does not capture any secrets, and you'd like to opt-out of secretness, you could use a [transformation](https://www.pulumi.com/docs/concepts/options/transformations/) to do so as a workaround. For example, calling `register_stack_transformation` before creating any of the dynamic resources:

```python
import pulumi

def unsecret_dynamic_providers(args: pulumi.ResourceTransformationArgs):
    if args.type_ == "pulumi-python:dynamic:Resource":
        args.props["__provider"] = pulumi.Output.unsecret(args.props["__provider"])
        return pulumi.ResourceTransformationResult(args.props, args.opts)
    return None

pulumi.runtime.register_stack_transformation(unsecret_dynamic_providers)
```

### PR DESCRIPTION
**Pull request by justinvp**
As of https://github.com/pulumi/pulumi/pull/13315 (which shipped in [v3.75.0](https://github.com/pulumi/pulumi/releases/tag/v3.75.0)), `__provider` (the serialized provider string) is _always_ set to a secret in the state. This can lead to poor performance when there are a lot of dynamic resources and the serialized provider does not actually have any secrets.

This change does two things:
1. Provides a way to opt-out of always serializing the provider as a secret.
2. Allows Outputs to be captured during serialization of the provider, which wasn't previously possible.

## 1. Opt-out of always serializing as secret

A new attribute, `serialize_as_secret_always`, can be set on a subclass of `ResourceProvider` to opt-out of always serializing the provider as a secret. If you know you don't have any secrets being serialized into the provider and want to avoid the encryption overhead, you can set this attribute to `False`.

```python
class MyProvider(ResourceProvider):
    serialize_as_secret_always = False

    def create(self, props):
        # Doesn't have any secrets that need encrypting
        ...
```

## 2. Allow Outputs to be captured

If you currently try to capture an `Output`, it fails when serializing the provider with:

```
TypeError: cannot pickle '_asyncio.Future' object
```

This change allows Outputs to be captured/serialized for dynamic providers, including secret Outputs.

This aligns Python dynamic providers with [Node.js](https://github.com/pulumi/pulumi/pull/13329).

```python
import pulumi
from pulumi.dynamic import CreateResult, Resource, ResourceProvider

config = pulumi.Config()
password = config.require_secret("password")

class SimpleProvider(ResourceProvider):
    def create(self, props):
        # Need to use `password.get()` to get the underlying value of the secret from within the serialized code.
        # This simulates using this as a credential to talk to an external system.
        return CreateResult("0", { "authenticated": "200" if password.get() == "s3cret" else "401" })

class SimpleResource(Resource):
    authenticated: pulumi.Output[str]

    def __init__(self, name):
        super().__init__(SimpleProvider(), name, { "authenticated": None })


r = SimpleResource("foo")
pulumi.export("out", r.authenticated)
```

Note: In the above example, we didn't have to specify `serialize_as_secret_always` since the default behavior is to always serialize the provider as a secret. If we wanted to, we could have specified `serialize_as_secret_always = False` and it still would have serialized the provider as a secret, since it captured `password` which is a secret Output. If `serialize_as_secret_always = False` was specified and no secrets were captured, then the provider would not be serialized as a secret.

We plan to recommend this approach to capturing secrets for both Node.js and Python dynamic providers after this change.

Fixes #15539

### CODE REVIEW

File: sdk/python/lib/pulumi/output.py [hunk 1]
```diff
@@ -139,6 +144,68 @@ def is_known(self) -> Awaitable[bool]:
...
+        self.resources = error("resources")
+        self.future = error("future")
+        self.is_known = error("is_known")
+        self.is_secret = error("is_secret")
+
+    def get(self) -> T_co:
+        """
+        Retrieves the underlying value of this Output.
+
+        This function is only callable in code that runs in the cloud post-deployment. At this
```
Comment (justinvp) [inline sdk/python/lib/pulumi/output.py:198]: I copied this text from Node.js SDK. Technically, we don't have full closure serialization in Python so the "in the cloud" part isn't really accurate. I could strip that part here. This is only really available to be called from a deserialized dynamic `ResourceProvider` subclass.

**Comment by justinvp** [comment]
After thinking about it more, I pushed a change the renames `auto_secret` to `serialize_as_secret_always` (and reverses the polarity). While it's a more verbose name, I think it better explains what it's doing.

If you're capturing secret Outputs, which this change enables and will be our recommended approach that we're going to document for both Node.js and Python dynamic providers, you won't need to set this attribute, because the provider will already always be a secret anyway (always making it a secret is the default behavior). If you wanted to specify `serialize_as_secret_always = False`, there's no harm in doing that; the provider would still be serialized as a secret since the captured Outputs are a secret. But you generally wouldn't have to set it.

However, if you know you aren't capturing any secrets and want to avoid the encrypt/decrypt performance overhead of always making it a secret, then you can specify `serialize_as_secret_always = False`.

File: tests/integration/integration_python_test.go [hunk 2]
```diff
@@ -488,6 +488,76 @@ func TestCustomResourceTypeNameDynamicPython(t *testing.T) {
...
+	integration.ProgramTest(t, &integration.ProgramTestOptions{
+		Dir: dir,
+		Dependencies: []string{
+			filepath.Join("..", "..", "sdk", "python", "env", "src"),
+		},
+		ExtraRuntimeValidation: func(t *testing.T, stack integration.RuntimeValidationStackInfo) {
+			randomVal = stack.Outputs["random_val"].(string)
+		},
+		EditDirs: []integration.EditDir{{
+			Dir:      filepath.Join(dir, "step1"),
```
Comment (julienp) [inline tests/integration/integration_python_test.go:506]: What s the empty edit here for? To check that nothing changes in a 2nd deployment?

**Review by julienp — APPROVED** [review]

File: tests/integration/integration_python_test.go (hunk 2 above)
Comment (justinvp) [reply tests/integration/integration_python_test.go:506]: Yes, I believe so. The first deployment gets the random value output and the 2nd confirms it didn't change. I just based this off the existing test:

https://github.com/pulumi/pulumi/blob/598b9837b6cd2aef5000f0bb0890bc009225efb5/tests/integration/integration_python_acceptance_test.go#L52-L97

File: sdk/python/lib/pulumi/dynamic/dynamic.py [hunk 3]
```diff
@@ -181,6 +182,14 @@ class ResourceProvider:
     whose CRUD operations are implemented inside your Python program.
     """
 
+    serialize_as_secret_always: bool = True
```
Comment (Frassle) [inline sdk/python/lib/pulumi/dynamic/dynamic.py:185]: This is better :)

**Pull request merged**